File: packages/java/anip-server/src/main/java/dev/anip/server/Storage.java
package dev.anip.server;

import dev.anip.core.ApprovalGrant;
import dev.anip.core.ApprovalRequest;
import dev.anip.core.AuditEntry;
import dev.anip.core.AuditFilters;
import dev.anip.core.Checkpoint;
import dev.anip.core.DelegationToken;

import java.io.Closeable;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * Abstract storage interface for ANIP server components.
 * Implementations handle tokens, audit log, checkpoints, and leases.
 */
public interface Storage extends Closeable {

    // --- Tokens ---

    /**
     * Persists a delegation token.
     */
    void storeToken(DelegationToken token) throws Exception;

    /**
     * Loads a delegation token by ID. Returns null if not found.
     */
    DelegationToken loadToken(String tokenId) throws Exception;

    // --- Audit ---

    /**
     * Atomically assigns sequence_number and computes previous_hash, then appends.
     * Returns the entry with sequence_number and previous_hash set.
     */
    AuditEntry appendAuditEntry(AuditEntry entry) throws Exception;

    /**
     * Queries audit entries with optional filters.
     */
    List<AuditEntry> queryAuditEntries(AuditFilters filters) throws Exception;

    /**
     * Returns the highest sequence_number, or 0 if empty.
     */
    int getMaxAuditSequence() throws Exception;

    /**
     * Returns audit entries with sequence_number between first and last (inclusive).
     */
    List<AuditEntry> getAuditEntriesRange(int first, int last) throws Exception;

    /**
     * Updates the signature on an existing audit entry (both column and JSON data blob).
     */
    void updateAuditSignature(int seqNum, String signature) throws Exception;

    // --- Checkpoints ---

    /**
     * Persists a checkpoint with its signature.
     */
    void storeCheckpoint(Checkpoint cp, String signature) throws Exception;

    /**
     * Returns checkpoints ordered by ID, limited by count.
     */
    List<Checkpoint> listCheckpoints(int limit) throws Exception;

    /**
     * Returns a checkpoint by its ID, or null if not found.
     */
    Checkpoint getCheckpointById(String id) throws Exception;

    // --- Retention ---

    /**
     * Deletes audit entries whose expires_at is before the given ISO timestamp.
     * Returns the number of deleted entries.
     */
    int deleteExpiredAuditEntries(String now) throws Exception;

    // --- Leases ---

    /**
     * Attempts to acquire an exclusive lease. Returns true if acquired.
     */
    boolean tryAcquireExclusive(String key, String holder, int ttlSeconds) throws Exception;

    /**
     * Releases an exclusive lease if held by the given holder.
     */
    void releaseExclusive(String key, String holder) throws Exception;

    /**
     * Attempts to acquire a leader lease for a background role.
     */
    boolean tryAcquireLeader(String role, String holder, int ttlSeconds) throws Exception;

    /**
     * Releases a leader lease if held by the given holder.
     */
    void releaseLeader(String role, String holder) throws Exception;

    // --- v0.23: ApprovalRequest + ApprovalGrant persistence ---

    /**
     * Persists an ApprovalRequest. Idempotent on approval_request_id when the
     * stored content matches; raises an exception when an existing row has
     * different content. v0.23. See SPEC.md §4.7.
     */
    void storeApprovalRequest(ApprovalRequest req) throws Exception;

    /**
     * Loads an ApprovalRequest by ID. Returns Optional.empty() if not found.
     */
    Optional<ApprovalRequest> getApprovalRequest(String approvalRequestId) throws Exception;

    /**
     * Atomically transitions a pending ApprovalRequest to "approved" and
     * persists the corresponding ApprovalGrant in a single transaction.
     * v0.23. SPEC.md §4.9 (Decision 0.9a).
     *
     * <p>Failure reasons (non-ok result):
     * <ul>
     *   <li>{@code approval_request_not_found} — no row matches the id.</li>
     *   <li>{@code approval_request_expired} — request expired before this txn.</li>
     *   <li>{@code approval_request_already_decided} — status was not "pending".</li>
     * </ul></p>
     *
     * @param approvalRequestId the request id
     * @param grant             the grant to persist (must be fully signed)
     * @param approver          approver principal map
     * @param decidedAtIso      timestamp recorded on the request
     * @param nowIso            current time used for the expiry CAS
     */
    ApprovalDecisionResult approveRequestAndStoreGrant(String approvalRequestId,
                                                       ApprovalGrant grant,
                                                       Map<String, Object> approver,
                                                       String decidedAtIso,
                                                       String nowIso) throws Exception;

    /**
     * Internal/test-only direct insert of an ApprovalGrant.
     */
    void storeGrant(ApprovalGrant grant) throws Exception;

    /**
     * Loads an ApprovalGrant by ID. Returns Optional.empty() if not found.
     */
    Optional<ApprovalGrant> getGrant(String grantId) throws Exception;

    /**
     * Atomically reserves one use of an unexpired grant by incrementing
     * use_count via a conditional UPDATE: WHERE grant_id=? AND
     * use_count &lt; max_uses AND expires_at &gt; now. v0.23. SPEC.md §4.8 Phase B.
     *
     * <p>On 0 affected rows the implementation re-reads the grant to
     * disambiguate {@code grant_not_found}, {@code grant_expired},
     * {@code grant_consumed}.</p>
     */
    GrantReservationResult tryReserveGrant(String grantId, String nowIso) throws Exception;
}

File: packages/go/server/storage.go
package server

import "github.com/anip-protocol/anip/packages/go/core"

// AuditFilters specifies optional filters for querying audit entries.
type AuditFilters struct {
	Capability         string `json:"capability,omitempty"`
	RootPrincipal      string `json:"root_principal,omitempty"`
	Since              string `json:"since,omitempty"`
	InvocationID       string `json:"invocation_id,omitempty"`
	ClientReferenceID  string `json:"client_reference_id,omitempty"`
	TaskID             string `json:"task_id,omitempty"`
	ParentInvocationID string `json:"parent_invocation_id,omitempty"`
	EventClass         string `json:"event_class,omitempty"`
	ApprovalRequestID  string `json:"approval_request_id,omitempty"`
	ApprovalGrantID    string `json:"approval_grant_id,omitempty"`
	Limit              int    `json:"limit,omitempty"`
}

// ApprovalDecisionResult is the outcome of an atomic
// approve_request_and_store_grant operation. v0.23 §4.9 / Decision 0.9a.
//
// On success, OK is true and Grant carries the persisted grant.
// On failure, OK is false and Reason is one of:
//   - "approval_request_not_found"
//   - "approval_request_expired"
//   - "approval_request_already_decided"
type ApprovalDecisionResult struct {
	OK     bool
	Reason string
	Grant  *core.ApprovalGrant
}

// GrantReservationResult is the outcome of an atomic try_reserve_grant
// operation. v0.23 §4.8 Phase B. Reason on failure is one of:
//   - "grant_not_found"
//   - "grant_expired"
//   - "grant_consumed"
type GrantReservationResult struct {
	OK     bool
	Reason string
	Grant  *core.ApprovalGrant
}

// Storage is the abstract storage interface for ANIP server components.
type Storage interface {
	// Tokens
	StoreToken(token *core.DelegationToken) error
	LoadToken(tokenID string) (*core.DelegationToken, error)

	// Audit
	AppendAuditEntry(entry *core.AuditEntry) (*core.AuditEntry, error)
	QueryAuditEntries(filters AuditFilters) ([]core.AuditEntry, error)
	GetMaxAuditSequence() (int, error)
	GetAuditEntriesRange(first, last int) ([]core.AuditEntry, error)
	UpdateAuditSignature(seqNum int, signature string) error

	// Checkpoints
	StoreCheckpoint(cp *core.Checkpoint, signature string) error
	ListCheckpoints(limit int) ([]core.Checkpoint, error)
	GetCheckpointByID(id string) (*core.Checkpoint, error)

	// Retention
	DeleteExpiredAuditEntries(now string) (int, error)

	// Leases (for horizontal scaling coordination)
	TryAcquireExclusive(key, holder string, ttlSeconds int) (bool, error)
	ReleaseExclusive(key, holder string) error
	TryAcquireLeader(role, holder string, ttlSeconds int) (bool, error)
	ReleaseLeader(role, holder string) error

	// v0.23: Approval requests + grants. See SPEC.md §4.7 / §4.8 / §4.9.
	// StoreApprovalRequest is idempotent on approval_request_id when content
	// is identical; conflicting re-store with the same id returns an error.
	StoreApprovalRequest(req *core.ApprovalRequest) error
	GetApprovalRequest(id string) (*core.ApprovalRequest, error)
	// ApproveRequestAndStoreGrant is the atomic security boundary for
	// issuance: conditional UPDATE of the request to status=approved AND
	// INSERT of the grant inside one transaction.
	ApproveRequestAndStoreGrant(
		approvalRequestID string,
		grant *core.ApprovalGrant,
		approver map[string]any,
		decidedAtIso string,
		nowIso string,
	) (ApprovalDecisionResult, error)
	// StoreGrant is internal/test-only: persists a grant without going
	// through the atomic approval pathway. Tests use it to seed fixtures.
	StoreGrant(grant *core.ApprovalGrant) error
	GetGrant(grantID string) (*core.ApprovalGrant, error)
	// TryReserveGrant atomically increments use_count if the grant is still
	// usable; returns the updated grant on success.
	TryReserveGrant(grantID string, nowIso string) (GrantReservationResult, error)

	// Close releases storage resources.
	Close() error
}
